### PR TITLE
AEIP-18 token resupply implementation

### LINKED ISSUE #1174: Implement AEIP2 token supply increase capability
### Is your feature request related to a problem?

_No response_

### Describe the solution you'd like

Since the AEIP18 we introduce a mechanism to support supply expendability for a token.
This specification introduces new fields and new verification to be applied during `token` transaction validation.

We should implement this specification to support this feature.

- [ ] We need to define/adapt the json schema relative to the token's minint specification using the pending validation to accept the `allow_mint` flag
- [ ] We need to introduce the `token_reference` field in the json schema to support reference a token tx to another one
- [ ] We need to ensure if the token transaction mentioned the AEIP18, the `allow_mint` field should be mentioned (required field)
- [ ] We need to ensure the `token_reference` can only be a token transaction
- [ ] We need to ensure the `token_reference` cannot be a transaction with a `token_reference` within
- [ ] We need to ensure the `token_reference` can only be applied to fungible tokens
- [ ] We need to ensure the `token_reference` can only be applied on a transaction with the `allow_mint: true` specified
- [ ] During validation, we need to use `token_reference` as address of the token for minted UTXO

### Additional context

_No response_

### Epic

_No response_

### PR DESCRIPTION
**Pull request by bchamagne**
# Description

Implement AEIP-18: token resupply
Fixes #1174 

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Manual scenario:
- create a fungible token
- create a resupply transaction
- utxos should have the same token_address

Screenshot_2023-08-01_16-04-41

plus a few unit tests

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Neylix — CHANGES_REQUESTED** [review]
I could be nice to also add test in ledger operation for token resuply

File: lib/archethic/mining/pending_transaction_validation.ex [hunk 1]
```diff
@@ -762,10 +767,28 @@ defmodule Archethic.Mining.PendingTransactionValidation do
...
-    with {:ok, json_token} <- Jason.decode(content),
-         :ok <- ExJsonSchema.Validator.validate(@token_schema, json_token),
-         %{
+  defp verify_token_transaction(tx = %Transaction{data: %TransactionData{content: content}}) do
+    case Jason.decode(content) do
+      {:error, _} ->
+        {:error, "Invalid token transaction - invalid JSON"}
+
+      {:ok, json_token} ->
+        case {ExJsonSchema.Validator.validate(@token_schema, json_token),
```
[the remark this answers is not in the document]
Comment (bchamagne) [reply lib/archethic/mining/pending_transaction_validation.ex:776]: I've done this while working on the issue #1175.

File: lib/archethic/mining/pending_transaction_validation.ex [hunk 2]
```diff
@@ -804,6 +824,53 @@ defmodule Archethic.Mining.PendingTransactionValidation do
     end
   end
 
+  defp verify_token_resupply(tx, %{"token_reference" => token_ref}) do
+    with {:ok, token_address} <- Base.decode16(token_ref, case: :mixed),
```
[the remark this answers is not in the document]
Comment (bchamagne) [reply lib/archethic/mining/pending_transaction_validation.ex:828]: I've done it in #1175 as well

File: lib/archethic/transaction_chain/transaction/validation_stamp/ledger_operations.ex [hunk 3]
```diff
@@ -140,6 +153,17 @@ defmodule Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperation
 
   def from_transaction(ops = %__MODULE__{}, %Transaction{}, _timestamp), do: ops
 
+  defp get_token_utxos(%{"token_reference" => token_ref, "supply" => supply}, address, timestamp) do
+    [
+      %UnspentOutput{
+        from: address,
+        amount: supply,
+        type: {:token, Base.decode16!(token_ref), 0},
```
[the remark this answers is not in the document]
Comment (bchamagne) [reply lib/archethic/transaction_chain/transaction/validation_stamp/ledger_operations.ex:161]: Done in #1175

File: priv/json-schemas/token-resupply.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        },
+        "aeip": {
+            "type": "array",
+            "description": "List of supported AEIPs",
+            "items": {
+                "type": "integer"
+            }
+        },
+        "token_reference": {
+            "type": "string",
```
[the remark this answers is not in the document]
Comment (bchamagne) [reply priv/json-schemas/token-resupply.json:19]: Done in #1175

File: lib/archethic/mining/pending_transaction_validation.ex [hunk 5]
```diff
@@ -804,6 +824,53 @@ defmodule Archethic.Mining.PendingTransactionValidation do
...
+         # verify same chain
+         {:ok, genesis_address} <- fetch_previous_tx_genesis_address(tx),
```
[the remark this answers is not in the document]
Comment (bchamagne) [reply lib/archethic/mining/pending_transaction_validation.ex:830]: I committed in #1175 https://github.com/archethic-foundation/archethic-node/compare/2566253d027672e6f561c9cdc43eeaff216003c5..e844e0f257893176a130585b7a93aa469b28a74b

**Pull request merged**